lib/error/try.h

```diff
@@ -0,0 +1,59 @@
+#ifndef LIB_ERROR_TRY_H_
+#define LIB_ERROR_TRY_H_
+
+#include "lib/base/paste.h"
+
+#define _TRY(expr, tmp)      \
+  {                          \
+    const auto tmp = (expr); \
+    if (IsError(tmp)) {      \
+      return tmp;            \
+    }                        \
+  }
+
+// Macro to evaluate an expression and immediately return an error status if
+// the expression returned an error. TRY() will work with any function that
+// returns a type implementing an overload of IsError().
+//
+// IsError() must be in the invoking namespace or discoverable using ADL.
+//
+// Example usage:
+//
+// Status Foo() {...}
+// Status Bar() {...}
+// Status Baz() {...}
+//
+// Status MyCompositionalFunction() {
+//   TRY(Foo());
+//   TRY(Bar());
+//   TRY(Baz());
+//   return ::error::kOkStatus;
+// }
+//
+// Compare to the same function without the TRY() macro:
+//
+// Status MyCompositionalFunction() {
+//   {
+//     const auto status = Foo();
+//     if (IsError(status)) {
+//       return status;
+//     }
+//   }
+//   {
+//     const auto status = Bar();
+//     if (IsError(status)) {
+//       return status;
+//     }
+//   }
+//   {
+//     const auto status = Baz();
+//     if (IsError(status)) {
+//       return status;
+//     }
+//   }
+//   return ::error::kOkStatus;
+// }
+//
+#define TRY(expr) _TRY(expr, PASTE(_try_tmp_, __COUNTER__))
```

Comment:
why do you need the PASTE and __COUNTER__? isn't the tmp variable in its own scope anyway?

(__COUNTER__ introduce a number of complexities in compilers - I recommend avoiding it whenever possible)

Reply:
I have never considered the compiler implications of COUNTER before. I'm glad you have LLVM experience! The counter here is really unnecessary, mostly used to mirror TRY_ASSIGN(). If I just use _try_tmp there is potential for aliasing against 'expr', however I don't have any sympathy for a user who declares their own '_try_tmp'.